File: transports/IConsumerConnection.go
// Package transports with the interface of a client transport connection
package transports

// IConsumerConnection defines the client interface of a consumer client connection
type IConsumerConnection interface {
	// ConnectWithClientCert connects to the server using a client certificate.
	// This authentication method is optional
	//ConnectWithClientCert(kp keys.IHiveKey, cert *tls.Certificate) (err error)

	// ConnectWithPassword connects to the messaging server using password authentication.
	// If a connection already exists it will be closed first.
	//
	// This returns a connection token that can be used with ConnectWithToken.
	//
	//  password is created when registering the user with the auth service.
	//
	// This authentication method must be supported by all transport implementations.
	ConnectWithPassword(password string) (newToken string, err error)

	// ConnectWithToken connects to the messaging server using an authentication token.
	//
	// If a connection already exists it will be closed first.
	//
	// and pub/private keys provided when creating an instance of the hub client.
	//  token is created by the auth service.
	//
	// This authentication method must be supported by all transport implementations.
	ConnectWithToken(token string) (newToken string, err error)

	// CreateKeyPair returns a new set of serialized public/private key pair.
	//  serializedKP contains the serialized public/private key pair
	//  pubKey contains the serialized public key to be shared
	//CreateKeyPair() (kp keys.IHiveKey)

	// Disconnect from the messaging server.
	// This retains the session and allows a reconnect with the session token.
	// See Logout for invalidating existing sessions
	Disconnect()

	// GetClientID returns the agent or user clientID for this hub client
	GetClientID() string

	// GetConnectionID returns the client's connection ID belonging to this endpoint
	GetConnectionID() string

	// GetProtocolType returns the (sub)protocol type implemented by this client
	// See the  ProtocolType constants defined above.
	// See also: https://www.w3.org/TR/wot-binding-templates/#protocol-bindings-table
	GetProtocolType() string

	// GetServerURL returns the schema://address:port of the server
	GetServerURL() string

	// InvokeAction sends an action request and waits for a response or until timeout.
	// This is short for SendRequest(wot.OpInvokeAction, ...)
	InvokeAction(dThingID, name string, input any, output any) error

	// IsConnected returns true when the connection is active
	IsConnected() bool

	// Logout of the hub and invalidate its session.
	// This will disconnect all existing connections from the hub and invalidate
	// all authentication tokens.
	// The client has to use login with password to reauthenticate.
	Logout() error

	// Ping the server and receive confirmation
	// Intended to check if the connection is alive.
	Ping() error

	// RefreshToken refreshes the authentication token
	// The resulting token can be used with 'ConnectWithToken'
	RefreshToken(oldToken string) (newToken string, err error)

	// SendRequest sends a request message.
	// wait is true to wait for a response or false to return immediately
	SendRequest(msg RequestMessage, wait bool) (ResponseMessage, error)

	// SetConnectHandler sets the callback for connection status changes
	// This replaces any previously set handler.
	SetConnectHandler(cb func(connected bool, err error))

	// SetNotificationHandler [consumer] sets the callback for receiving notifications.
	// This replaces any previously set handler.
	SetNotificationHandler(cb NotificationHandler)

	// SetResponseHandler [consumer] sets the callback for receiving unhandled responses
	// to requests. If a request is sent with 'sync' set to true then SendRequest
	// will handle the response instead.
	//
	// This replaces any previously set handler.
	SetResponseHandler(cb ResponseHandler)

	// ObserveProperty [consumer] observes changes to a property.
	//
	// The protocol binding handles this as per its specification
	// This is a convenience method and short for
	//  SendRequest(wot.ObserveProperty,thingID,name,nil)
	// name is the property to subscribe to or "" for all properties
	ObserveProperty(thingID string, name string) error
	// UnobserveProperty removes a previous observe of a property
	UnobserveProperty(thingID string, name string) error
	// Subscribe to one or all events of a thing
	// name is the event to subscribe to or "" for all events
	Subscribe(thingID string, name string) error
	// Unsubscribe from previous subscription
	Unsubscribe(thingID string, name string) error

	// WriteProperty submits a request to modify a property
	// This is short for SendRequest(wot.OpWriteProperty, ...)
	//	thingID is the thing whose property to write
	//	name is the property affordance name
	//	input is the value to write as per affordance dataschema
	//	async is true to return after submitting the request or false to wait for a response
	WriteProperty(thingID, name string, input any, async bool) error
}

File: libjs/src/messaging/IConsumerConnection.ts
import { RequestMessage, ResponseMessage} from "./Messages.ts";

export enum ConnectionStatus {
    Connected = "connected",
    Connecting = "connecting",
    ConnectFailed  = "connectFailed",
    Disconnected = "disconnected",
    // Unauthorized login name or password
    Unauthorized = "unauthorized"
}


export type ConnectionHandler = (status: ConnectionStatus)=>void;
export type RequestHandler = (msg: RequestMessage)=>ResponseMessage|null;
export type ResponseHandler = (resp: ResponseMessage)=>void;

// IConsumerConnection defines the interface of the consumer facing protocol binding.
export default interface IConsumerConnection {

    // ConnectWithPassword connects to the hub using password authentication.
    // @param password is created when registering the user with the auth service.
    // This returns an authentication token that can be used in refresh and connectWithToken.
    connectWithPassword(password: string): Promise<string>

    // ConnectWithToken connects to the messaging server using an authentication token
    // and pub/private keys provided when creating an instance of the hub client.
    // @param token is created by the auth service.
    connectWithToken(token: string): Promise<string>


    // Disconnect from the messaging server.
    disconnect(): void;

    // getStatus returns the current transport connection status
    // getStatus(): TransportStatus

    // invokeAction sends an action request and waits for a response or until timeout.
    // This is a simple helper that uses sendRequest(wot.OpInvokeAction, ...)
    //
    //	@param dThingID the digital twin ID for whom the action is intended
    //	@param name is the action or method name of the action to invoke
    //	@param input data to publish in native format as per TD
    //
    invokeAction(dThingID: string, name: string, input: any): Promise<ResponseMessage>

    // RefreshToken refreshes the authentication token
    // The resulting token can be used with 'ConnectWithJWT'
    refreshToken(): Promise<string>

    // Rpc sends a request message and waits for a response.
    //
    //  operation to request
    //	dThingID is the digital twin ID of the service providing the RPC method
    //	key is the ID of the RPC method as described in the service TD action affordance
    //	input is the value containing the input arguments sent
    //
    // This returns the data or throws an error if failed
    rpc(operation: string, dThingID: string, key: string, input: any): Promise<any>

    // sendRequest sends a request message and returns a response on completion or error.
    // if no response is received within a time period a timeout response is returned.
    sendRequest(msg: RequestMessage): Promise<ResponseMessage>

    // set handler that is notified of changes in connection status and an error in
    // case of an  unintentional disconnect.
    // 
    // This handler is intended for updating presentation of the connection status.
    // Do not call connectXyz() in this handler, as a reconnect attempt will be made 
    // after a short delay. If a connection is re-established then the onConnect
    // handler will be invoked.
    //
    //  status contains the connection status, eg connected, disconnected
    //
    // If a reconnect is to take place with a different password or token then 
    // call disconnect(), followed by connectWithXyz().
    setConnectHandler(handler: (status: ConnectionStatus) => void): void


    // Set the progress handler
    // This replaces any previously set handler.
    setResponseHandler(handler: ResponseHandler):void

    // Subscribe adds a subscription for events from the given ThingID.
    //
    //  dThingID is the digital twin ID of the Thing to subscribe to. ""  for any
    //	name is the event name to subscribe to or "" for all events, "" for any
    subscribe(dThingID: string, name:string): void

// Unsubscribe removes a previous event subscription.
// No more events or requests will be received after Unsubscribe.
//     unsubscribe(dThingID: string, key: string): void;

    // writeProperty consumer requests a configuration change
    //  @param dThingID is the digitwin thingID is provided by the directory
    //	@param name ID of the property
    //	@param payload to publish in native format as per TD
    writeProperty(dThingID: string, name: string, payload: any):void
}
